innerprod/fullysec/paillier.go

```diff
@@ -0,0 +1,237 @@
+	n       *big.Int   // a big integer, a product of two safe primes
+	nSquare *big.Int   // n^2 a modulus for computations
+	boundX  *big.Int   // a bound on the entries of the input vector
+	boundY  *big.Int   // a bound on the entries of the inner product vector
+	sigma   *big.Float // the standard deviation for the sampling a secret key
+	lambda  int        // safety parameter
+	g       *big.Int   // generator of the 2n-th residues subgroup of Z_n^2*
+}
+
+// Paillier represents a scheme based on the Paillier encryption.
+type Paillier struct {
+	Params *paillerParams
+}
+
+// NewPaillier configures a new instance of the scheme.
+// It accepts the length of input vectors l, security parameter lambda,
+// the bit length of prime numbers (giving security to the scheme, it
+// should be such that factoring two primes with such a bit length takes
+// at least 2^lambda operations), and boundX and boundY by which
+// coordinates of input vectors and inner product vectors are bounded.
+//
+// It returns an error in the case the scheme could not be properly
+// configured, or if the precondition boundX, boundY < (n / l)^(1/2)
+// is not satisfied.
+func NewPaillier(l, lambda, bitLen int, boundX, boundY *big.Int) (*Paillier, error) {
+	// generate two safe primes
+	p, err := emmy.GetSafePrime(bitLen)
+	if err != nil {
+		return nil, err
+	}
+
+	q, err := emmy.GetSafePrime(bitLen)
+	if err != nil {
+		return nil, err
+	}
+
+	// calculate n = p * q
+	n := new(big.Int).Mul(p, q)
+
+	// calculate n^2
+	nSquare := new(big.Int).Mul(n, n)
+
+	// check if the parameters of the scheme are compatible,
+	// i.e. security parameter should be big enough that
+	// the generated n is much greater than l and the bounds
+	xSquareL := new(big.Int).Mul(boundX, boundX)
+	xSquareL.Mul(xSquareL, big.NewInt(int64(l)))
+	ySquareL := new(big.Int).Mul(boundY, boundY)
+	ySquareL.Mul(ySquareL, big.NewInt(int64(l)))
+	if n.Cmp(xSquareL) < 1 {
+		return nil, fmt.Errorf("parameters generation failed," +
+			"boundX and l too big for bitLen")
+	}
+	if n.Cmp(ySquareL) < 1 {
+		return nil, fmt.Errorf("parameters generation failed," +
+			"boundX and l too big for bitLen")
+	}
+
+	// generate a generator for the 2n-th residues subgroup of Z_n^2*
+	gPrime, err := rand.Int(rand.Reader, nSquare)
+	if err != nil {
+		return nil, err
+	}
+	g := new(big.Int).Exp(gPrime, n, nSquare)
+	g.Exp(g, big.NewInt(2), nSquare)
+
+	// check if generated g is invertible, which should be the case except with
+	// negligible probability
+	check := new(big.Int).ModInverse(g, nSquare)
+	if check == nil {
+		return nil, fmt.Errorf("parameters generation failed," +
+			"unexpected event of generator g is not invertible")
+	}
+
+	// calculate sigma
+	nTo5 := new(big.Int).Exp(n, big.NewInt(5), nil)
+	sigma := new(big.Float).SetInt(nTo5)
+	sigma.Mul(sigma, big.NewFloat(float64(lambda)))
+	sigma.Sqrt(sigma)
+	sigma.Add(sigma, big.NewFloat(2))
+	// make it an integer for faster sampling with sample_double
```

Comment:
I assume `sample_double` is a reference to `sample.NormalDouble` sampler? If so, please update the comment.